opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,595 @@
+          result->total_elapsed_time = this->now() - action_start_time_;
+          result->error_code = FollowObject::Result::TIMEOUT;
+          result->num_retries = num_retries_;
+          publishZeroVelocity();
+          following_action_server_->succeeded_current(result);
+          return;
+        }
+
+        // Approach the object using control law
+        if (approachObject(object_pose)) {
+          // We have reached the object, maintain position
+          RCLCPP_INFO_THROTTLE(
+            get_logger(), *get_clock(), 1000,
+            "Reached object. Stopping until goal is moved again.");
+          publishFollowingFeedback(FollowObject::Feedback::STOPPING);
+          publishZeroVelocity();
+          continue;
+        }
+
+        // Cancelled, preempted, or shutting down (recoverable errors throw FollowingException)
+        result->total_elapsed_time = this->now() - action_start_time_;
+        publishZeroVelocity();
+        following_action_server_->terminate_all(result);
+        return;
+      } catch (opennav_following::FollowingException & e) {
+        if (++num_retries_ > max_retries_) {
+          RCLCPP_ERROR(get_logger(), "Failed to follow, all retries have been used");
+          throw;
+        }
+        RCLCPP_WARN(get_logger(), "Following failed, will retry: %s", e.what());
+
+        // Perform an in-place rotation to find the object again
+        if (!rotateToObject(object_pose)) {
+          // Cancelled, preempted, or shutting down
+          publishZeroVelocity();
+          following_action_server_->terminate_all(result);
+          return;
+        }
+        RCLCPP_INFO(get_logger(), "Rotated to find object again");
+      }
+    }
+  } catch (const tf2::TransformException & e) {
+    RCLCPP_ERROR(get_logger(), "Transform error: %s", e.what());
+    result->error_code = FollowObject::Result::TF_ERROR;
+  } catch (opennav_following::ObjectNotValid & e) {
+    RCLCPP_ERROR(get_logger(), "%s", e.what());
+    result->error_code = FollowObject::Result::OBJECT_NOT_VALID;
+  } catch (opennav_following::FailedToDetectObject & e) {
+    RCLCPP_ERROR(get_logger(), "%s", e.what());
+    result->error_code = FollowObject::Result::FAILED_TO_DETECT_OBJECT;
+  } catch (opennav_following::FailedToControl & e) {
+    RCLCPP_ERROR(get_logger(), "%s", e.what());
+    result->error_code = FollowObject::Result::FAILED_TO_CONTROL;
+  } catch (opennav_following::FollowingException & e) {
+    RCLCPP_ERROR(get_logger(), "%s", e.what());
+    result->error_code = FollowObject::Result::UNKNOWN;
+  } catch (std::exception & e) {
+    RCLCPP_ERROR(get_logger(), "%s", e.what());
+    result->error_code = FollowObject::Result::UNKNOWN;
+  }
+
+  // Stop the robot and report
+  result->num_retries = num_retries_;
+  publishZeroVelocity();
+  following_action_server_->terminate_current(result);
+}
+
+void FollowingServer::doInitialPerception(geometry_msgs::msg::PoseStamped & dock_pose)
```

Comment:
This method doesn't do perception, it just looks for a value in a topic updating the pose. I think it needs to be renamed, that caused me confusion in my initial readthrough

Reply:
Actually from reading this too, I don't think this method is actually required. `approachObject` immediately checks on getting the updated pose as well